### PR TITLE
Add file metadata columns support for spark parquet

### PR DESCRIPTION
**Pull request by gaoyangxiaozhu**
Spark support people query file metdata as `file_size`, `file_name`, `file_path`, `file_modified_time`, `file_block_start` etc for Hive tables as seperated file metadata column.  Checking this for all file metadata supported query by spark https://github.com/apache/spark/blob/081c7a7947a47bf0b2bfd478abdd4b78a1db3ddb/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala#L183C2-L193C56.

The PR extends the `HiveSplit` with a new parameter `metadaColumns` to let upsteram computing engine as spark to pass the initialized const file metadata columns (if have) to velox connector split when constructed, it use to fix file metadata columns null issue when intergrating with Velox using Spark.

1) It let HiveSplit can get file metadata info from spark
2) The values is populated by by SplitReader into TableScanOperator output.

Checking this issue https://github.com/facebookincubator/velox/issues/8173 for details context.

It is also a dependency of Gluten repository PR https://github.com/oap-project/gluten/pull/3870

### CODE REVIEW

File: velox/connectors/hive/HiveDataSource.cpp [hunk 1]
```diff
@@ -782,7 +788,8 @@ std::shared_ptr<common::ScanSpec> HiveDataSource::makeScanSpec(
     // $bucket column. This filter is redundant and needs to be removed.
     // TODO Remove this check when Presto is fixed to not specify a filter
     // on $path and $bucket column.
-    if (auto name = pair.first.toString(); name == kPath || name == kBucket) {
+    if (auto name = pair.first.toString();
+        name == kPath || name == kBucket || metadataColumns.count(name) != 0) {
```
Comment (Yuhta) [inline velox/connectors/hive/HiveDataSource.cpp:792]: Can you handle & remove such filters in Gluten?  This should be handled in split scheduling not inside table scan.  We will also fix it in presto_cpp at some point.

Comment (gaoyangxiaozhu) [reply velox/connectors/hive/HiveDataSource.cpp]: @Yuhta sure, make sense.

Comment (gaoyangxiaozhu) [reply velox/connectors/hive/HiveDataSource.cpp]: and apologize for so late to resolve the comment due to busy those week with internal project release work.

**Comment by gaoyangxiaozhu** [comment]
@majetideepak  also help for review

**Comment by aditi-pandit** [comment]
@gaoyangxiaozhu : Have couple of high-level comments. There are several missing pieces :

i) A TableScan output operator test like https://github.com/facebookincubator/velox/pull/8800/files#r1498654870. This would introduce the wiring for metadata columns in HiveConnector TestBase classes.
ii) The metadata columns can be used as filters in Presto queries as well. I would imagine Spark would allow this as well. The code should be enhanced for this.
iii) What is the difference between synthesized and metadata columns ? Its not clear to me.

It might be simpler to change https://github.com/facebookincubator/velox/pull/8800 to use metadata_columns parameters for HiveSplit instead. wdyt ?

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu : Have couple of high-level comments. There are several missing pieces :
> 
> i) A TableScan output operator test like https://github.com/facebookincubator/velox/pull/8800/files#r1498654870. This would introduce the wiring for metadata columns in HiveConnector TestBase classes. ii) The metadata columns can be used as filters in Presto queries as well. I would imagine Spark would allow this as well. The code should be enhanced for this. iii) What is the difference between synthesized and metadata columns ? Its not clear to me.
> 
> It might be simpler to change #8800 to use metadata_columns parameters for HiveSplit instead. wdyt ?

thanks @aditi-pandit 

for i)  Yes, a test is needed and I saw your PR already have it.
for ii) Actually I add filter part support and just remove it temporily with this commit https://github.com/facebookincubator/velox/pull/7880/commits/df5e279f644ef8f669c90ab0faf51ef5170389be, but we can added it back if it still need for making both presto and spark work.

image

for iii) I actually before don't know what's the synthesized means and also i don't found anywhere code path use it, so i just added a KMetadata to easily mark it is metadata column, but we can still use synthesized it should be ok, I'd like metadata naming due to I don't understand what synthesized means.

 Sure @aditi-pandit , it's OK if you can update your PR to move using metadata_columns parameters and remove any hard code checking for know it is metadata or not.  For filter part just reference change 

image
image

**Comment by gaoyangxiaozhu** [comment]
closed with @aditi-pandit 's PR is merged do the same thing.

**Pull request closed without merging**